pcsx2/Recording/InputRecording.cpp

```diff
@@ -91,58 +91,57 @@ void InputRecording::ControllerInterrupt(u8& data, u8& port, u16& bufCount, u8 b
 
 	if (bufCount == 1)
 		fInterruptFrame = data == READ_DATA_AND_VIBRATE_FIRST_BYTE;
-	else if (bufCount == 2 && buf[bufCount] != READ_DATA_AND_VIBRATE_SECOND_BYTE)
-		fInterruptFrame = false;
-
-	// We do not want to record or save the first two bytes in the data returned from the PAD plugin
-	else if (fInterruptFrame && bufCount >= 3 && frameCounter >= 0 && frameCounter < INT_MAX)
+	else if (bufCount == 2)
+	{
+		if (buf[bufCount] != READ_DATA_AND_VIBRATE_SECOND_BYTE)
+			fInterruptFrame = false;
+	}
+	else if (fInterruptFrame)
 	{
-		u8& bufVal = buf[bufCount];
-		const u16 bufIndex = bufCount - 3;
-
 		if (state == InputRecordingMode::Replaying)
 		{
-			u8 tmp = 0;
-			if (inputRecordingData.ReadKeyBuffer(tmp, frameCounter, port, bufIndex))
+			if (frameCounter >= 0 && frameCounter < INT_MAX)
 			{
-				// Overwrite value originally provided by the PAD plugin
-				bufVal = tmp;
+				u8& bufVal = buf[bufCount];
+				const u16 bufIndex = bufCount - 3;
+				if (!inputRecordingData.ReadKeyBuffer(bufVal, frameCounter, port, bufIndex))
+					recordingConLog(wxString::Format(L"[REC]: Failed to read input data at frame %d.\n", frameCounter));
+
+				// Update controller data state for future VirtualPad / logging usage.
+				padData[port]->UpdateControllerData(bufIndex, bufVal);
+				
+				if (virtualPads[port]->IsShown())
+					virtualPads[port]->UpdateControllerData(bufIndex, padData[port]);
 			}
 		}
-
-		// Update controller data state for future VirtualPad / logging usage.
-		padData[port]->UpdateControllerData(bufIndex, bufVal);
-
-		if (virtualPads[port] &&
-			virtualPads[port]->IsShown() &&
-			virtualPads[port]->UpdateControllerData(bufIndex, padData[port]) &&
-			state != InputRecordingMode::Replaying)
-		{
-			// If the VirtualPad updated the PadData, we have to update the buffer
-			// before committing it to the recording / sending it to the game
-			// - Do not do this if we are in replay mode!
-			bufVal = padData[port]->PollControllerData(bufIndex);
-		}
-
-		// If we have reached the end of the pad data, log it out
-		if (bufIndex == PadData::END_INDEX_CONTROLLER_BUFFER)
-		{
-			padData[port]->LogPadData(port);
-			// As well as re-render the virtual pad UI, if applicable
-			// - Don't render if it's minimized
-			if (virtualPads[port] && virtualPads[port]->IsShown() && !virtualPads[port]->IsIconized())
-				virtualPads[port]->Redraw();
-		}
-
-		// Finally, commit the byte to the movie file if we are recording
-		if (state == InputRecordingMode::Recording)
+		else
 		{
-			if (incrementUndo)
+			const u16 bufIndex = bufCount - 3;
```

Comment:
this value is calculated twice identically, can probably be calculated earlier, once

Reply:
Not necessarily. If, for example, `state == InputRecordingMode::Replaying` *but* you loaded a savestate that is set before the first frame of the recording, bufIndex would not get used until you reach frame 0. Having it calculated beforehand would be pointless.